keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,567 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The [kube-storage-version-migrator] supports migrating existing resources to a new storage version but it does not modify a CRD's `status.storedVersions` field post successful migration. Instead, the cluster administrator is responsible for manually modifying the CRD's `status.storedVersions`. The amount of work this design decision generates for Cluster Administrators will only increase as CRDs are adopted and new versions emerge.
+
+Historically, the [kube-storage-version-migrator] has not managed modifying the `status.storedVersion` field because the CRD's storedVersion could change mid-migration unbenownst to the migrator due to slow/failed watches. 
+
+This KEP proposes a solution that would allow the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` after a successful migration.
+
+[kube-storage-version-migrator]: https://github.com/kubernetes-sigs/kube-storage-version-migrator
+
+## Motivation
+
+It is desirable for the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` array after a successful migration rather than relying on a cluster admin to manually perform the same steps.
+
+### Goals
+
+- Allow the [kube-storage-version-migrator] to update a CRD's `status.storedVersions`  with the list of stored versions after a successful migration.
+- Ensure that the kube-storage-version-migrator has successfully done its job before removing a stored version from the status of the CRD.
+
+### Non-Goals
+
+- Modify a CRD's `status.storedVersions` field if the migration is not successful.
+
+## Proposal
+
+In order to safely modify a CRD's `status.storedVersions` field, the [kube-storage-version-migrator] would need to prevent changes to the CRD during a migration. CRD's could be marked as migrating using an annotation and changes to the CRD could be prevented by means of an [Validating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+
+The validating admission webhook would need to:
+
+- Intercept all Update and Patch operations against CRDs that include the `ksvm.k8s.io/migrating=true` annotation.
+- Allow `/status` requests from the [kube-storage-version-migrator] service account.
+- Allow setting or removing the migrating annotation by the [kube-storage-version-migrator] service account.
+- Prevent other changes to the CRD.
```

Comment:
>The only interest ksvm can have is to avoid that the storage version is changed by another actor while migrating. Isn't it enough to prevent only that?

I agree that the primarily requirement is to prevent changes to the storedVersion. I am concerned about removing the `ksvm.k8s.io/migrating=true` annotation as it simplified / reduced the cost of identifying which CRDs the webhook needs to intercept. Without this annotation, it seems like the webhook would need to:
1. Intercept all CRD patch/update requests
2. Check if a [storageVersionMigration](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/master/manifests/storage_migration_crd.yaml) CR exists for the CRD
3. If CRD is being migrated, prevent changes to the storedVersion.

Unless there is an alternative approach that I'm not considering, it seems like we'd introduce a bit more compute when updating/patching crds. This increased load might be negligible based on how infrequently CRDs change, but I wanted to have the discussion.

Reply:
I am not against the annotation. I am just saying we don't have to block every update, but can greatly restrict the set of disallowed updates to those changing the storage version.

Reply:
Understood, I will update the webhook to rely on the presence of the annotation to prevent changes to the stored version.

Reply:
To the storage version, i.e. `spec.versions[*].storage`.

Reply:
Updated the description of the proposed webhook based on this feedback.